x/marker/keeper/send_restrictions.go

```diff
@@ -0,0 +1,140 @@
+package keeper
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	attrTypes "github.com/provenance-io/provenance/x/attribute/types"
+	"github.com/provenance-io/provenance/x/marker/types"
+)
+
+const AddressHasAccessKey = "address_has_access"
+
+func (k Keeper) AllowMarkerSend(ctx sdk.Context, from, to, denom string) error {
+	markerAddr := types.MustGetMarkerAddress(denom)
+	marker, err := k.GetMarker(ctx, markerAddr)
+	if err != nil {
+		return err
+	}
+	if marker == nil { // this should only occur in tests
+		return nil
+	}
+
+	if marker.GetMarkerType() != types.MarkerType_RestrictedCoin {
+		return nil
+	}
+
+	caller, err := sdk.AccAddressFromBech32(from)
+	if err != nil {
+		return err
+	}
+
+	hasAccess, err := GetAddressHasAccess(ctx)
+	if err != nil {
+		return err
+	}
+
+	// if the marker has authority it is allowed to send to receiver without checking of attributes
+	if hasAccess || marker.AddressHasAccess(caller, types.Access_Transfer) || caller.Equals(k.markerModuleAddr) {
+		return nil
+	}
+
+	if len(marker.GetRequiredAttributes()) == 0 {
+		return fmt.Errorf("%s does not have transfer permissions", caller.String())
+	}
+	contains, err := k.ContainsRequiredAttributes(ctx, marker.GetRequiredAttributes(), to)
+	if err != nil {
+		return err
+	}
```

Comment:
The `hasAccess` check should happen right off the bat (so that all the lookup and processing is skipped).
The `caller.Equals(k.markerModuleAddr)` check should happen next since it can be done and doesn't require any lookups.

Then, the `return fmt.Errorf("%s does not have transfer permissions", caller.String())` should be inside an if block checking for transfer permissions.